src/protocol/slot.rs

```diff
@@ -0,0 +1,186 @@
+use std::io::{Read, Seek, SeekFrom, Write};
+
+use crate::nbt::Compound;
+use crate::protocol::{Decode, Encode, VarInt};
+
+/// Represents a slot in an inventory.
+#[derive(Clone, Default, Debug)]
+pub enum Slot {
+    #[default]
+    Empty,
+    Present {
+        item_id: VarInt,
+        item_count: u8,
+        nbt: Option<Compound>,
+    },
+}
+
+impl Slot {
+    /// Constructs a new slot.
+    pub const fn new(item_id: VarInt, item_count: u8, nbt: Option<Compound>) -> Self {
+        Self::Present {
+            item_id,
+            item_count,
+            nbt,
+        }
+    }
+
+    /// Constructs an empty slot.
+    pub const fn empty() -> Self {
+        Self::Empty
+    }
+
+    /// Returns `true` if there is an item present.
+    pub const fn is_present(&self) -> bool {
+        match self {
+            Slot::Empty => false,
+            Slot::Present { .. } => true,
+        }
+    }
+
+    /// Gets the item id.
+    ///
+    /// If the slot is empty, then `None` is returned.
+    pub fn item_id(&self) -> Option<&VarInt> {
+        match self {
+            Slot::Empty => None,
+            Slot::Present { item_id, .. } => Some(item_id),
+        }
+    }
+
+    /// Gets the item count.
+    ///
+    /// If the slot is empty, then `None` is returned.
+    pub fn item_count(&self) -> Option<&u8> {
+        match self {
+            Slot::Empty => None,
+            Slot::Present { item_count, .. } => Some(item_count),
+        }
+    }
+
+    /// Gets the item's nbt data.
+    ///
+    /// If the slot is empty or there is no nbt data, then `None` is returned.
+    pub fn nbt(&self) -> Option<&Compound> {
+        match self {
+            Slot::Empty => None,
+            Slot::Present { nbt, .. } => nbt.as_ref(),
+        }
+    }
+}
+
+impl Encode for Slot {
+    fn encode(&self, w: &mut impl Write) -> anyhow::Result<()> {
+        match self {
+            Slot::Empty => false.encode(w),
+            Slot::Present {
+                item_id,
+                item_count,
+                nbt,
+            } => {
+                true.encode(w)?;
+                item_id.encode(w)?;
+                item_count.encode(w)?;
+                match &nbt {
+                    Some(n) => n.encode(w),
+                    None => 0u8.encode(w),
+                }
+            }
+        }
+    }
+}
+
+impl Decode for Slot {
+    fn decode(r: &mut (impl Read + Seek)) -> anyhow::Result<Self> {
+        let present = bool::decode(r)?;
+        if !present {
+            return Ok(Slot::empty());
+        }
+        Ok(Slot::new(
+            VarInt::decode(r)?,
+            u8::decode(r)?,
+            match u8::decode(r)? {
+                0 => None,
+                _ => {
+                    r.seek(SeekFrom::Current(-1))?;
+                    Some(Compound::decode(r)?)
+                }
+            },
```

Comment:
We can do this better with `&mut &[u8]`. Consider:

```rust
let some_bytes = &mut &[1, 2, 3, 4, 5, 6, 7];

// Is there a zero byte at index zero?
if some_bytes.get(0) == Some(&0) {
    some_bytes.read_u8()?;
    // No NBT data.
} else {
    // There is NBT data.
}
```

Reply:
Oh, I see now. That is a lot simpler than the `Read + Seek`. Should the `Encode` trait signature also be updated to use `&mut &[u8]`?

Reply:
The analog for `Write` is actually `&mut [u8]` but you can leave it.